TypeScript Node.js implementation

Oxygen588 opened the pull request.

Oxygen588 committed:
Ts

Oxygen588 committed:
Fix&Style guidelines

Oxygen588 committed:
Formatting.

Fixing the indentation and spacing.

JOT85 commented on typescript/redis_work_queue/index.ts:
```diff
@@ -0,0 +1,231 @@
+  async complete(db: Redis, item: typeof Item) {
+    // Mark an item as complete.
+    await db.del(KeyPrefix(this.leaseKey).of(item.Id()));
+    await db.lrem(this.processingKey, 0, item.Id());
```
These can be pipelined and sent together, rather than waiting for one and then the other!

Oxygen588 committed:
Removed unnecessary code.

Removed all unnecessary code, made concat easier to work with.

JOT85 commented on typescript/redis_work_queue/index.ts:
```diff
@@ -0,0 +1,231 @@
+  }
+
+  async complete(db: Redis, item: typeof Item) {
+    // Mark an item as complete.
```
This should return a boolean, indicating if this was the first complete call to be made.

Also, you haven't removed it from the processing queue here!

Oxygen588 committed:
Fixed complete method.

Fixed the removing from processing Queue,  added the delete commands to pipeline.

Oxygen588 committed:
Update index.ts

JOT85 commented on typescript/redis_work_queue/index.ts:
```diff
@@ -0,0 +1,251 @@
+    }
+
+    // Make sure the item id is a string
+    if (Buffer.isBuffer(maybeItemId)) {
+      itemId = maybeItemId.toString('utf-8');
+    } else if (typeof maybeItemId === 'string') {
```
According to https://redis.github.io/ioredis/classes/Redis.html#rpoplpush, the result will always be a string.

JOT85 commented on typescript/redis_work_queue/index.ts:
```diff
@@ -0,0 +1,251 @@
+    }
+
+    // If we got an item, fetch the associated data.
+    let data: Buffer | string | null = await db.get(KeyPrefix(this.itemDataKey).of(itemId));
```
What's the behaviour of `get`, what happens if the value can't be converted to a string?

Oxygen588 committed:
Lease,Complete, Cleaning methods fix, created test file for typescript.

The complete method had issues related to related the way the delete key was created. // Fixed
LightClean method did not have await at one of the redis commands which resulted in items left in processing queue.
Created Test file for TypeScript and added everything to the run_tests.sh.

JOT85 commented on typescript/redis_work_queue/index.ts:
```diff
@@ -0,0 +1,248 @@
+
+    if (Buffer.isBuffer(maybeItemId)) {
+        maybeItemId = maybeItemId.toString('utf-8');
+    }
```
This can't be a buffer!

JOT85 commented on typescript/redis_work_queue/Item.ts:
```diff
@@ -0,0 +1,83 @@
+    return new Item(JSON.stringify(data), id);
+  }
+
+  Data() {
+    // Get the data associated with this item.
+    return this.data;
+  }
```
There's no need for this method

JOT85 commented on typescript/redis_work_queue/Item.ts:
```diff
@@ -0,0 +1,83 @@
+  Id(): string {
+    // Get the ID of the item.
+    return this.id;
+  }
```
There's no need for this method

JOT85 commented on typescript/redis_work_queue/index.ts:
```diff
@@ -0,0 +1,248 @@
+    return db.llen(this.mainQueueKey);
+  }
+
+  processing(db: Redis) {
```
Type annotation?

Oxygen588 committed:
Added TypeScript in README, Got read of not necessary code, Fixed comments, File name changes.

1. Added TypeScript into the Readme, will need updated when the docs website is up with the right links.
2. the Id and Data method within Item Class as they can be now directly accessed, removed parse static method as it is not necessary.
3.The comments have been modified so they work properly with TypeDocs, as well as comments have been moved from inside methods to the right place so they appear properly in the docs.
4.The index file name has been modified to a more appropriate name "WorkQueue".

JOT85 committed:
Add type annotations to `WorkQueue.processing`

JOT85 committed:
lrange doesn't return buffers

JOT85 committed:
Remove extra spacing

JOT85 committed:
Don't catch the error if parsing fails!

JOT85 committed:
Make typescript a dev dependency

JOT85 committed:
Add typescript license

Oxygen588 committed:
Created TypeScript readme, Fixed Package.json, got rid of nested folders.

1. Created Typescript readme and created showcase on how this can be used.
2. Fixed the package.json to allow the publication on npm.
3. Got rid of the nested folders "redis_work_queue" now it is directly.

JOT85 commented on typescript/src/Item.ts:
```diff
@@ -0,0 +1,83 @@
+   * @param {ItemData} loaded The data needed to be converted to a item.
+   * @returns {Item} a new Item instance loaded with item data.
+   */
+  static fromDict(loaded: ItemData): Item {
```
Perhaps we should rename this to `fromObject`?

JOT85 commented on typescript/src/WorkQueue.ts:
```diff
@@ -0,0 +1,270 @@
+    db: Redis,
+    leaseSecs: number,
+    block = true,
+    timeout = 1
```
Super pedantic: I think this would look nicer with a training comma, does the formatter remove that?

Oxygen588 replied:
Yes, it does remove it.

JOT85 commented on typescript/src/WorkQueue.ts:
```diff
@@ -0,0 +1,270 @@
+    leaseSecs: number,
+    block = true,
+    timeout = 1
+  ): Promise<typeof Item> {
```
(I don't know a whole lot of typescript so there's probably a good reason I'm not aware of) Is there a reason this is `typeof Item` and not just `Item`?

Oxygen588 replied:
Typeof gets the type of the class Item where if we would use Item directly it would be a instance of Item.

JOT85 replied:
Are you sure? In JavaScript, the expression `Item` wouldn't be an instance of it, it would be the class itself (the constructor function). So `typeof Item` wouldn't be the `Item` type. I bet it's something more general here, too.

Further to this point, TypeScript allows you to return `null`, which isn't an `Item`. You shouldn't be allowed to return `null` if the return type must be an `Item`.

I don't think `typeof Item` is what you think it is. I think it's just a general object, or something else. Does it allow you to return `{}`, or `123` and not complain?

Oxygen588 replied:
You are right, the way I was importing Item did not allow me to use the class as a type.

I fixed it, will push after.

JOT85 commented on typescript/src/WorkQueue.ts:
```diff
@@ -0,0 +1,270 @@
+      maybeItemId = await db.brpoplpush(
+        this.mainQueueKey,
+        this.processingKey,
+        timeout
```
Trailing comma? Unless specified not to in the style guide.

JOT85 commented on typescript/src/WorkQueue.ts:
```diff
@@ -0,0 +1,270 @@
+   * Process Explenation:
+   * If the lease key is not present for an item (it expired or was never created because the client crashed before creating it), then move the item back to the main queue so others can work on it.
+   * While working on an item, we store it in the cleaning list. If we ever crash, we come back and check these items.
+   */
+  async clean(db: Redis) {
+    const processing: Array<string> = await db.lrange(
+      this.processingKey,
+      0,
+      -1
+    );
+
+    for (const itemId of processing) {
+      if (!this.leaseExists(db, itemId)) {
+        await db.lpush(this.cleaningKey, itemId);
+        const removed = Number(db.lrem(this.processingKey, 0, itemId));
+        if (removed > 0) {
+          await db.lpush(this.processingKey, 0, itemId);
+        }
+        await db.lrem(this.cleaningKey, 0, itemId);
+      }
+    }
+
+    const forgot: Array<string> = await db.lrange(this.cleaningKey, 0, -1);
+    for (const itemId of forgot) {
+      const leaseExists: boolean = await this.leaseExists(db, itemId);
+      if (
+        !leaseExists &&
+        (await db.lpos(this.mainQueueKey, itemId)) == null &&
+        (await db.lpos(this.processingKey, itemId)) == null
+      ) {
+        /**
+         * FIXME: this introduces a race
+         * maybe not anymore
+         * no, it still does, what if the job has been completed?
+         */
+        await db.lpush(this.mainQueueKey, itemId);
+      }
+      await db.lrem(this.cleaningKey, 0, itemId);
+    }
+  }
```
This is a duplicate of `lightClean`?

Oxygen588 replied:
Yes it is, I will remove it.

JOT85 committed:
Use `Node.js` instead of `NodeJs`

JOT85 committed:
Uncomment C# tests

JOT85 committed:
tests: typescript: fix formatting

JOT85 committed:
Don't excute the pipeline in `WorkQueue.addToPipeline`!!

JOT85 committed:
ts: convert lease exists function to async and reorder exports

JOT85 committed:
ts: convert WorkQueue.addItem to async

JOT85 committed:
ts: update package-lock.json

Oxygen588 committed:
DS_Store remove,Editor Config, formatting.

Removed unnecessary DS_Store files.
Included Editor Config.
Formated TypeScript Readme.

Oxygen588 committed:
Merge branch 'JavaScript-Implementation' of https://github.com/Oxygen588/redis-work-queue into JavaScript-Implementation

Oxygen588 committed:
Readme Fix, Removed duplicate clean method

Readme docs fixed, added header for docs.
Deleted the duplicated cleaning method.

Oxygen588 committed:
Readme changes

Added NPM report badge.

Oxygen588 committed:
Update README.md

Oxygen588 committed:
Rewritten the tests bash file, Rewritten parts ofpython-spawner-and-cleaner, removed typeof from all files.

Rewritten test bash file, now it allows you to test specific languages, created Instructions on how to use within README file.
python-spawner-and-cleaner: Parts of it ware Rewritten to make it to allow it to set which jobs directly from bash file.
Removed all type of within WorkQueue.

JOT85 commented on tests/job-spawner-and-cleaner.py:
```diff
@@ -4,22 +4,28 @@
-#dotnet_queue = WorkQueue(KeyPrefix("dotnet_jobs"))
-shared_queue = WorkQueue(KeyPrefix("shared_jobs"))
+
+queue_list.append("shared_jobs")
+queue_list_names = []
+
+for index in range(len(queue_list)):
+    queue_list_names.append(queue_list[index])
+    queue_list[index]=WorkQueue(KeyPrefix(queue_list[index]))
+
```
We could have used `.insert(0, item)` to prepend the items.

However, there's no need for that. I think it's much cleaner and way more readable to keep `shared_queue` as a separate variable and only put the language specific queues in the list.

JOT85 commented on tests/job-spawner-and-cleaner.py:
```diff
@@ -4,22 +4,28 @@
 if len(sys.argv) < 2:
     raise Exception("first command line argument must be redis host")
-db = redis.Redis(host=sys.argv[1])
+db = redis.Redis(host=sys.argv[1].replace(":6379",""))
```
Why not split the argument into host and port and pass these both? With this, if I pass `localhost:4567` it won't connect to the right port!

JOT85 commented on tests/job-spawner-and-cleaner.py:
```diff
@@ -28,128 +34,140 @@
-        [(a+b, a*b) for a in [19, 23] for b in expecting_dotnet] + \
-        [(a+b, a*b) for a in [17, 21] for b in expecting_typeScript]
+expecting_dict_config={
+"python_jobs": {
+    "expecting share":[13,17],
+    "expected": [(n * 3)%256 for n in range(0, 256*3)],
+     "result_name":"results:python:"
+},
+"rust_jobs": {
+    "expecting share":[3,5],
+    "expected": [(n * 7)%256 for n in range(0, 256*3)],
+     "result_name":"results:rust:"
+},
+"go_jobs": {
+    "expecting share":[7,11],
+    "expected":[(n * 5)%256 for n in range(0, 256*3)],
+     "result_name":"results:go:"
+},
+"typeScript_jobs": {
+    "expecting share":[17,21],
+    "expected":[(n * 17)%256 for n in range(0, 256*3)],
+    "result_name":"results:typeScript:"
+},
+"dotnet_jobs": {
+    "expecting share":[17,21],
+    "expected":[(n * 11)%256 for n in range(0, 256*3)],
+    "result_name":"results:dotnet:"
+}
```
Indent these!!

JOT85 commented on tests/python-tests.py:
```diff
@@ -9,7 +9,7 @@
 if len(sys.argv) < 2:
     raise Exception("first command line argument must be redis host")
-db = redis.Redis(host=sys.argv[1])
+db = redis.Redis(host=sys.argv[1].replace(":6379",""))
```
Same port thing as above, pass the port if one is received, don't ignore it!

JOT85 commented on typescript/src/Item.ts:
```diff
@@ -76,7 +76,7 @@ export class Item {
     } else {
       jsonString = this.data;
-    }
+    } 
```
No need for that space! Check your `git diff` before you commit!

JOT85 committed:
tests: use iterator for doom counter

JOT85 committed:
tests: use if expression for doom counter

Oxygen588 committed:
Fixed Redis connection in python,Indentation, removed empty space.

Fixed Redis connection in python by dividing the port and host properly.

JOT85 commented on tests/job-spawner-and-cleaner.py:
```diff
@@ -13,13 +13,18 @@
     raise Exception("first command line argument must be redis host")
-db = redis.Redis(host=sys.argv[1].replace(":6379",""))
+
+host = sys.argv[1].split(":")
+
+if len(host) == 2:
+    db = redis.Redis(host=host[0], port=host[1])
+else:
+    db = redis.Redis(host=host[0])
```
What if I pass `localhost:123:456`?

Also, `host[1]` will be a string, is that allowed to be passed as `port`?

JOT85 commented on tests/job-spawner-and-cleaner.py:
```diff
@@ -34,66 +39,70 @@
         # Spawn 256 initial jobs in each queue
-        for queue in range(len(queue_list)-1):
+        for queue in range(len(queue_list)):
             queue_list[queue].add_item(db, Item(bytes([counter])))
```
No need to index, just do `for queue in `queue_list`.

JOT85 commented on tests/job-spawner-and-cleaner.py:
```diff
@@ -13,13 +13,18 @@
+shared_jobs = WorkQueue(KeyPrefix("shared_jobs"))
 queue_list_names = []
 
 for index in range(len(queue_list)):
```
Why copy `queue_list` to `queue_list_names` and overwrite `queue_list` when you could just make one new list and leave the names unmodified?

JOT85 commented on tests/job-spawner-and-cleaner.py:
```diff
@@ -34,66 +39,70 @@
             n = n % 256
-            for queue in range(len(queue_list)-1):
+            for queue in range(len(queue_list)):
                 queue_list[queue].add_item(db, Item(bytes([n])))
```
Again, no need to index. Use `for queue in queue_list`

JOT85 commented on tests/job-spawner-and-cleaner.py:
```diff
@@ -34,66 +39,70 @@
         revived = True
         for n in range(0, 256):
-            for queue in range(len(queue_list)-1):
+            for queue in range(len(queue_list)):
```
No need to index!

JOT85 commented on tests/job-spawner-and-cleaner.py:
```diff
@@ -34,66 +39,70 @@
     doom_counter = doom_counter + 1 if all(map(
         lambda queue: queue.queue_len(db) == 0 and queue.processing(db) == 0, queue_list,
-    )) else 0
+    )) and shared_jobs.processing(db) == 0 and shared_jobs.queue_len(db) == 0  else 0
```
We could instead iterate over `queue_list + [shared_queue]` to remove the duplicated code.

Oxygen588 committed:
Update job-spawner-and-cleaner.py

Fixed required changes.

JOT85 commented on tests/job-spawner-and-cleaner.py:
```diff
@@ -8,28 +8,25 @@
-else:
-    db = redis.Redis(host=host[0])
+
+db = redis.Redis(host=host[0], port=int(host[1]) or 6379)
```
This definitely doesn't work if the port isn't passed!

JOT85 commented on tests/job-spawner-and-cleaner.py:
```diff
@@ -8,28 +8,25 @@
-    queue_list_names.append(queue_list[index])
-    queue_list[index]=WorkQueue(KeyPrefix(queue_list[index]))
+for name in queue_list_names:
+    queue_list.append(WorkQueue(KeyPrefix(name)))
```
This could be replaced with:
```python
queue_list = list(map(
    lambda name: WorkQueue(KeyPrefix(name)),
    queue_list_names,
))
```

JOT85 replied:
Maybe I'm too prone to functional things, it's fine how it is.

Oxygen588 committed:
Update job-spawner-and-cleaner.py

Fixed connection logics.
modified queue_list creation.

JOT85 commented on tests/python-tests.py:
```diff
@@ -9,7 +9,13 @@
+host = sys.argv[1].split(":")
+
+if len(host) == 2:
+    db = redis.Redis(host=host[0], port=host[1])
```
Does `port` accept a string?

Oxygen588 committed:
Update python-tests.py

Fixed redis connection logics for python test.

JOT85 committed:
Tidy up tests

JOT85 committed:
tests: correct all tests instructions

JOT85 committed:
Formatting

JOT85 committed:
Tidy up lots of little things in the TypeScript implementation

JOT85 committed:
tests: neaten up test scripts

JOT85 committed:
tests: update run-test.sh help

JOT85 committed:
Rename TypeScript -> Node.js
